README.txt
ai64_extractor - C64 file extractor

Copyright (c) 2004-2021 Ferenc 'Lion/Kempelen' Veres
lion@c64.rulez.org https://lion.xaraya.hu

WHAT IS THIS?

ai64 allows you to convert complete directory structures containing
c64 programs into IDE64 compatible copy of the whole structure. e.g.
try running it on a complete copy of Digital Dungeon FTP Archive, before
burning a CD for IDE64 usage. Most of the archive will be converted to 
.prg and .d64 files, which are very easy to use on C64!

SYSTEM REQUIREMENTS

PHP-CLI
    Which makes it possible to run PHP programs from the command line.
Linux/Unix
    Tested on Linux only. Windows compatible filename generation is now 
    supported, Windows platform support is planned (never tested).
unzip
    Command line unzip to decompress the files.
unrar
    Command line unrar to decompress the files.
tar
    Command line tar to decompress the files.
gzip
    Command line gzip to decompress the files.
zip2disk
    Zipcode to d64 converter by Marko Makela. 
    https://c64.rulez.org/pub/c64/other-OS/Unix
cbmconvert
    Marko Makela's converter masterpiece to handle all the c64 file formats
    (tested with v2.1.2).
    ftp://ftp.zimmers.net/pub/cbm/crossplatform/converters/unix
c1541
    Command line utility, part of Vice emulator (tested with v2.2).
    https://vice-emu.sourceforge.io/
d64list [Not needed if you use Vice c1541]
    This is a little C program I made years ago, included in ai64 download.
    https://c64.rulez.org/pub/c64/other-OS/Unix/

WHAT DOES IT DO?

The program will process a complete directory of C64 files
convert all files to c64 usable format and copy them into a destination
directory structure. Thus from all your ZIP, TGZ, etc files, you will
get a big directory directly usable on your IDE64 or via PCLink cable,
SD card readers, pendrive compatible cards like 1541 Ultimate, or on
Ultimate 64 or a C64 Emulator.

Currently the following formats are handled (anything can be nested, 
e.g. a single prg in a d64 in a zip):

C64 files:
D64: extracted if 1 file only and BAM is valid. Kept as D64 otherwise.
     Hi-score files (if identified) are ignored if there is only 1 more file.
     Some BAM validation messages in names/header are taken into consideration.
T64: extracted to files (if there is a single file named "FILE" is inside,
     the name of the original t64 is used when saving it to prg.)
P00: extracted to normal files
PRG: copied as is

Compressed files:
ZIP: extracted and re-processed
TAR: extracted and re-processed
GZ : extracted and re-processed
TGZ: extracted and re-processed
RAR: extracted and re-processed
LNX: converted to d64 and re-processed
ZIPCODE: (1!,2!..) converted to d64 and re-processed

Not copied (not usable on C64):
TXT (also extensionless README, 00INDEX)
DIZ, ME, NFO
COM, EXE
AVI, MPG, MPEG
PDF, DOC, DJVU
PNG, JPG, JPEG, GIF
DB
INI
C and any one letter extension (causes error with FuseCFS)
DEL (D64 extraction dirt)
DIR (IDE64 dir type, cannot be copied)
LNK (cannot be copied to IDE64)
REL (cannot be copied to IDE64)

Files starting with dot are not copied (Unix hidden files, FTP site messages).

Not listed files: copied as is.

Support is missing for: BZIP2, LHA (lzh).

All the stored files are saved with characters which are readable on C64.
Filenames are converted to 16 + extension, IDE64 can handle this length 
(requires 0.9x or higher IDEDOS). Already existing files are not 
overwritten, the new file will get "-1", "-2".. index, thus running the
program twice will create all destination files twice.

When the whole directory structure conversion is finished, ai64 will make
another process. It will rearrange all the directories which contain 
more than 300 files, to make sure there are no more than 300 files in a
directory (easier to handle on c64, MAN, etc). The created subdirs will
be called "ai100-X", "ai200-X" and so on, where "X" means the first word
of the first program in that directory. E.g. "ai300-blackjack".

INSTALLATION

Edit ai64 executable file if necessary: 
1. Customize the location of your php interpreter in the first line. 
2. Set the location of your d64list program or set it to empty string to 
   bypass analyzing d64 contents.
3. Make sure the -t tmp_dir is far from any important locations, because the
   program will run thousands of "rm -rf" commands inside it (honestly, I 
   create a "lion2" user just to run the program, you never know... 
   especially while developing it.) See ramdisk tip below.
4. Customize the location of "cbmconvert" executable as well.
5. Make sure the other decompress executables are on your path.
6. There are some other configurations in the source's head.

USAGE

1. Collect the c64 stuff what you want to convert into a directory structure
2. Run "ai64.php sourcedirname destinationdirname"

ai64.php [options] original_dir destination_dir

    -s path/name    Skip to this source file before staring processing
                    (Use this to continue after something went wrong)
    -x ,            Use ',' as file extension separator (default '.')
    -n 100          Number of maximum files per folder (default 300)
    -v              Verbose, list successfully processed files
    -V              Super-verbose, also list archives while processing them
    -w              Force windows compatible file naming (remove more chars)
    -u              Enable unicode chars like ↑ | (remove otherwise)
    -t path         Temp dir for extractions (default '/tmp/[USER].ai64/')
    -e err_handling Error handling, either 'ignore' (default), 'ask' or 'halt'.

If ".php" is not registered to your PHP interpreter, you may need to type 
"php ai64.php" instead, assuming php.exe is on your PATH.

There is no warranty of any kind. So I advice again, to create a temporary
user which cannot write your home directory, and run the whole conversion by
that user!

Examples

ai64.php -x , -v -w c64/Demos c64stuff

  Converts with comma as extension separator, windows compatible file names.
  Useful when copying to HD using FuseCFS (https://singularcrew.hu/idedos/).

ai64.php downloadedwares coolwares

  Converts with dot as separator. Linux file naming conventions.
  Useful when burning a CD for use with IDE64.

TIPS

1. Using Ramdisk or TmpFS

Since the program creates and deletes a LOT OF TEMPORARY FILES while 
uncomressing the archives, it is a good idea to choose a temporary dir 
(configred in the script) in RAM. 

Using tmpfs (recommended)

sudo mount -t tmpfs none /mnt/rd
sudo chown lion2:lion2 /mnt/rd

Using ramdisk (more complicated)

Usually Linux distros have ramdisk by default, to find it see:

ls /dev/ram*
dmesg | grep -i ramdisk

The second one also displays the size, what you can configure with the 
"ramdisk_size=64000" (64MB) kernel parameter at boot time (bootloader config).
I use the following small shell script as root, before converting:

mke2fs -m 0 /dev/ram0
mount /dev/ram0 /mnt/rd
chown lion2:lion2 /mnt/rd

2. Logging the errors

There is no option in the program to log the error messages, but it is very
simple using standard Unix tool, "tee" (-a is for append, if needed). This
will also help to find where some buggy files originated from and fix
errors in your original archive dirs.

ai64.php orig_dir dest_dir 2>&1 | tee -a errorlog.txt

LICENSE

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

CONTACT

Feedback and patches are welcome on lion@c64.rulez.org.
Source code on GitHub: https://github.com/lionkmp/ai64_extractor/
Find my converted c64 downloads on my blog: https://lion.xaraya.hu/



CHANGELOG.txt
******************************
** ai64_exctrator ChangeLog **
******************************

  1.6 (2021/04/04)
        Change default to 300 files/dir.
        Fixes to unicode support (use mb_* routines where necessary).
        Fix errors in natsort() natural sorting implementation.
        Fix an error in invoking unrar.
        Unicode "\" is now incomatible with FuseCFS, removed.
        Add -t switch to set temp dir.
        Add -e switch to control error handling behavior.

  1.5 (2019/09/09)
  		Update to PHP 7.x
  		Unicode support for up arrow chars and others.
		Skip PC document types: doc, pdf, jpg, etc
		Skip extensionless readme files (00index, readme)
		Skip IDEDOS special types (they show "invalid argument" on fusecfs)
		they are: del, dir, lnk, rel, and one letter extensions like .c
		Fix unrar problem when unrar asked something.
		More fixes to dir-rearrange.
		Implement -V super-verbose mode.
		(2015/06/13)
		Update to current PHP.
		Update to current IDEDOS (Remove chars like ^, will need Unicode)
		Add -n NUM option to choose number of files per dir.
		Other improvements to dir-rearrange.
  
  1.4 (2010/03/24/lion):
		Change to GetOpt library, argument format changed
		Support -x for setting comma as seaprator (-x ,)
		Support -v for verbose
		Support -w for forcing windows file naming
		Support -s to "skip to file" (moved from third argument)
		Allow only ASCII chars in output filenames
		Restrict output filename chars when used on Windows
		Add "prg" if no extension exists on save
		Add "noname" if no name left after character filters
		Rename dirs to fit length on IDE64, similar to filenames
		Dir name collision was not cheched in generating filenames
		Remove repeated dots and spaces from names also trim these
		Relicensed under GPL V3
		No comma in output filenames
      
  1.3 (2006/01/02/lion):
  		Support for c1541 (VICE) to list d64 files (default from now)
		(Illegal track&sector errors won't stop with c1541 because
		it returns OK exit code as of version 4.00.)

  1.2 (2005/12/27/lion):
  		PHP 5.x warnings of uninitialized variables fixed
		Code formatting changed from K&R to new
		Rearrange dirnames contain first file part (eg. "ai300-blackjack")
		Errors from d64list were ignored. Not from now.

  1.1 (2005/01/23/lion):
  		Adding some error handling, fatal error quits (mkdir,opendir)
		Added a config (ERRORHALT) for archiver error quits
  		TAR, GZ, TGZ, RAR, LNX, ZIPCODE support added
		D64 now can ignore separators and hi-scores to some degree
		Config variables are now uppercase
		Cleaning up tmp dir avoids copying old files again
		Checking for required tools, quit with error if missing
		Handling invalid chars in filenames and invalid extensions

  1.0 (2004/10/23/lion):
  		First version:
		System architecture design
		PRG, D64, T64, ZIP, P00 support
		Max 100 file rearranging



make-demo.php
#!/usr/bin/php
<?php
/* 
This script makes the example output that is
lined from the ai64 site.
*/

if($argc != 2 || !is_dir($argv[1]))
{
	echo "Usage: make-demo.php dirname >demo.html";
	exit(1);
}

show_header();
list_dir($argv[1]);

function list_dir($dir)
{
	$dh = opendir($dir);
	if ($dh === false)
	{
		echo("Cannot open dir for processing: $dir\n");
		exit(1);
	}
	unset($dirs_here);
	unset($files_here);
	
	// Collect files and dirs into two arrays

	while (($file = readdir($dh)) !== false) 
	{
		if ($file != '.' && $file != '..')
		{
			$fullname = "$dir/$file";

			if (is_dir($fullname))
			{
				$dirs_here[] = $file;
			}
			elseif (is_file($fullname))
			{
				$files_here[] = $file;
			}
		}
	}
	closedir($dh);

	// Process all subdirs here
	if (!empty($dirs_here) && count($dirs_here) > 0)
	{
		sort($dirs_here);
		foreach ($dirs_here as $newdir) 
		{
			echo("<fieldset>\n");
			echo("<legend>$dir/$newdir</legend>\n");
			list_dir("$dir/$newdir");
			echo("</fieldset>\n");
		}
	}

	// Process all files here
	if (!empty($files_here) && count($files_here) > 0) 
	{
		sort($files_here);
		foreach ($files_here as $file) 
		{
			echo("<div>$file</div>\n");
		}
	}

}
show_footer();

function show_header()
{
?>
<html>
<head>
<title>ai64 demo result</title>
<style type="text/css">
body 
{
	font-family: sans-serif;
	background: black;
	color: white;
}
a
{
	color: yellow;
}
fieldset div
{
	float: left;
}
legend
{
	border: solid 1px yellow;
	padding: 5px;
	margin: 0px;
}
fieldset
{
	margin: 5px;
	padding: 10px;
}
</style>
<link rel="stylesheet" type="text/css" href="col4.css" title="Four columns" media="screen" />
<link rel="alternate stylesheet" type="text/css" href="col2.css" title="Two columns" media="screen" />
<link rel="alternate stylesheet" type="text/css" href="col3.css" title="Three columns" media="screen" />
<link rel="alternate stylesheet" type="text/css" href="col5.css" title="Five columns" media="screen" />

</head>
<body>
<h1>ai64 demo</h1>
<p>
<a href="http://lion.xaraya.hu/news/70">ai64 batch extractor</a> creates this output from
<a href="ftp://c64.rulez.org/pub/c64/Scene/Old">games on c64.rulez.org FTP</a>.
</p>

<p>You can download archives like this from the 
<a href="http://c64.rulez.org/xlfiles">C64 XLFiles</a> website.</p>
<p>Primary aim is to use with IDE64 or other hard disk solutions or with C64 emulators. 
Most of the files in this collection are converted directly to PRG or D64, regardless
of the original format. Some files have been broken by the conversion, minor issue. 
To convert other archive, get the program from the link above.</p>
<p>Customize 2-3-4-5 columns view to your monitor: 
use your browser's "View-&gt;Stylesheet" menu.</p>
<p>&copy; Lion/Kempelen 2005-2010</p>
<?php
}

function show_footer()
{
?>
</body>
</html>
<?php
}
?>


d64list.c
/*
D64 lister and file extractor V0.2  (c) Lion of Chromance 1998

You can use this program to list your .D64 emulator files,
and extract them into binary c64 files on Unix.

Using: d64list [-lcef] [disk_image.d64] [disk_image2.d64] ...
  -l = use lowercase charset (like CBM+SHIFT)
  -c = start in copy mode. You can accept the default filename
       by pressing enter or type a new one
  -e = don't add extensions (.prg,.seq,etc) to the default filenames
       in copy mode
  -f = print diskimage filenames

If there is no input filename given, the program uses the standard
input, reads the first d64 file.

lion@c64.rulez.org
*/

#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <errno.h>
#define EMPTY '.'
#define TRUE 1
#define FALSE 0
#define D64SIZE (256*683)

char *filetypes[] = {"DEL","SEQ","PRG","USR","REL","???","???","???" };
int fext=TRUE;               /* file extension for default copy names */
int lowercase = FALSE;       /* lowercase chars for listing */
int copymode = FALSE;        /* -c = copy mode */
int namepr = FALSE;          /* print d64 filenames too */
unsigned char d64[683][256]; /* the disk image */

/*** loads the d64 file into an array (from stdin if fname==NULL) ***/
int d64load(char *fname) {
 int i,fil,alldatas;
 char temp[1];

 if (fname!=NULL) {                   /* opening the file, if not stdin */
     if ((fil = open(fname,O_RDONLY)) < 0) {
        printf("d64list: cannot open %s\n",fname);
        close(fil); return(FALSE);
        }
 	i=read(fil,d64,D64SIZE);                          /* loading the datas */
 	if (i!=D64SIZE) {
		printf("d64list: error reading %s (too small for a .D64 file)\n",fname);
		close(fil); return(FALSE);
		}
     i=read(fil,temp,1);                /* checking if file is too long */
     if (i!=0) {
        printf("d64list: error reading %s (too big for a .D64 file)\n",fname);
        close(fil); return(FALSE);
        }
     close(fil);
    }
 else {                        /* loading from stdin, continously trying */
 	alldatas=0;
 	while (alldatas!=D64SIZE) {
		i=read(STDIN_FILENO,d64+alldatas,D64SIZE-alldatas);
		alldatas+=i;
		printf("%d %d\n",alldatas,i);
		perror("d64list");
		if (i<0) exit(1);
		}
	}

 return(TRUE);
}

/*** calculate the index in the d64[] array of a sector and track ***/
int d64seek(int trk,int sec) {
  int pos=0;   /*  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18
                  19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 */
  static int sectors[]={
                  21,21,21,21,21,21,21,21,21,21,21,21,21,21,21,21,21,19,
                  19,19,19,19,19,19,18,18,18,18,18,18,17,17,17,17,17};

  if (trk>35 || sectors[trk-1]<sec) {           /* correct request? */
    fprintf(stderr,"67,ILLEGAL TRACK OR SECTOR,%d,%d\n",trk,sec);
    exit(1);
    }

  trk--;                                          /* tracks = 0-34 */
  for (;trk>0;trk--)                                   /* counting */
	pos += sectors[trk-1];
  pos += sec;
  return (pos);
}

/*** convert one c64 char into ASCII code for listing ***/
int d64print(int betu) {
  int nbetu;
  if (!lowercase) {
    if (betu<32) nbetu=EMPTY;
    else if (betu<65) nbetu=betu;
    else if (betu<91) nbetu=betu;
    else if (betu==91) nbetu=91;
    else if (betu==92) nbetu=EMPTY;
    else if (betu==93) nbetu=93;
    else if (betu==94) nbetu=EMPTY;
    else if (betu==95) nbetu=EMPTY;
    else if (betu==96) nbetu=45;
	else nbetu=EMPTY;
	}
  else {
	if (betu<32) nbetu=EMPTY;
	else if (betu<65) nbetu=betu;
	else if (betu<91) nbetu=betu+32;
	else if (betu==91) nbetu=91;
	else if (betu==92) nbetu=EMPTY;
	else if (betu==93) nbetu=93;
	else if (betu==94) nbetu=EMPTY;
	else if (betu==95) nbetu=EMPTY;
	else if (betu==96) nbetu=45;
	else if (betu<123) nbetu=betu-32;
	else if (betu<193) nbetu=EMPTY;
	else if (betu<219) nbetu=betu-128;
	else nbetu=EMPTY;
	}
  return(nbetu);
}

/*** generate a filename from a the ASCII coded c64 filename ***/
void d64conv(char *oname,char *nname,int tipus) {
  int first=TRUE;       /* first printable character? */
  int wordt=0;
  int i;
  i=0;
  do {
    if (first && isprint(*oname)) first=FALSE;  /* no output until the */
    if (!first) {                               /* first printable char*/
		if (isalnum(*oname))
            if (wordt!=1) {                       /* word's first char */
                *nname++=toupper(*oname);
                wordt=1;                                    /* in word */
			}
            else *nname++=tolower(*oname);
        else if (*oname=='-') {                        /* '-'? sotring */
                *nname++='-';
            wordt=2;                                /* like a new word */
			}
        else {                      /* no normal chars, but we were in */
            if (wordt==1) *nname++='_';    /* a nice word, printing '_'*/
            wordt=2;                          /* like a new word again */
		}
	}
  }
  while (*++oname!='\0');                                 /* go to end */
  while (*--nname=='_') ;               /* strip trailling underscores */
  if (fext) {                       /* adding file extension if wanted */
	*++nname='.';
	for (i=0;*(filetypes[tipus]+i);i++)
		*++nname=tolower(*(filetypes[tipus]+i));
	}
   *++nname='\0';
}

/*** save the file with fnev filename from the given track and sector ***/
int f_save(char *fnev,unsigned char trk,unsigned char sec) {
 int siz,sblokk,sfil;
 mode_t jogok;

 jogok=S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH;
 if ((sfil=open(fnev,O_WRONLY|O_CREAT,jogok))>=0) {
	do {
		sblokk=d64seek(trk,sec);
        siz=d64[sblokk][0]?254:d64[sblokk][1]-1;        /* last block? */
		if (write(sfil,d64[sblokk]+2,siz)!=siz) return(1);
		trk=d64[sblokk][0]; sec=d64[sblokk][1];
		}
	while (trk!=0);
	close(sfil);
	} else {
	printf("d64list:error opening file %s",fnev);
	return(1);
	}
  return(0);
}

/*** listing one d64 file, main function ***/
void d64list(char *filenev) {
  int i,j,k,blokk,ntrk,nsec,filecount;
  int cr=TRUE;
  char attrib,sor[BUFSIZ],nname[BUFSIZ],oname[BUFSIZ];

  if (d64load(filenev)) {

  printf("\n");
  if (namepr) printf("%s:\n",filenev);                    /* d64 filename */
  blokk=d64seek(18,0);                              /* disk header and ID */
  i=0; k=d64[blokk][144]; if (k==160) k=32;
  while (i<23) {
	oname[i]=d64print(k);
	k=d64[blokk][144+(++i)]; if (k==160) k=32;
	}
  oname[i]='\0'; oname[16]='"';
  printf("0 \"%s",oname);

  /*  ntrk=d64[blokk][0]; nsec=d64[blokk][1]; not following link! */

  ntrk=18; nsec=1; filecount=0;
  do {                                                    /* listing the dir */
	blokk=d64seek(ntrk,nsec);
	for (j=0; j < 256 && filecount<144 ;j+=32) {
		attrib=d64[blokk][j+2];
		if (attrib != 0) {
            if (cr) printf("\n");         /* user sent LF coz saved a file? */
			printf("%-5d\"",d64[blokk][j+31]*256+d64[blokk][j+30]); /* blks */
			i=0; k=d64[blokk][j+5];                             /* filename */
			while (i<16 && k!=160) {
				oname[i]=d64print(k);
				k=d64[blokk][j+5+(++i)];
				}
			oname[i]='\0';
			printf("%s\"",oname);
            for (;i<16;i++) putchar(' ');               /* padding filename */
			if (attrib & 0x80) putchar(' '); else putchar('*');   /* opened */
			if (lowercase)                                      /* filetype */
				for (i=0;i<3;i++)
					putchar(tolower(*(filetypes[attrib & 0x07]+i)));
				else printf("%s",filetypes[attrib & 0x07]);
			if (attrib & 0x40) putchar(' '); else putchar('<'); /* proteced */
			cr=TRUE;
			if (copymode && (attrib & 0x07) &&                 /* copy mode */
                (attrib & 0x07) < 4 && (attrib & 0x80)) {   /* normal file? */
				do {
                    printf(" (y/n)");                      /* want to save? */
					fgets(sor,100,stdin);
					}
				while (*sor!='n' && *sor!='y');
				if (*sor=='y') {
                    d64conv(oname,nname,attrib & 0x07);     /* default name */
					printf("Destination (%s):",nname);
					fgets(sor,100,stdin); *(sor+strlen(sor)-1)='\0';
					i=f_save((*sor=='\0')?nname:sor,d64[blokk][j+3],d64[blokk][j+4]);
					cr=FALSE;
					}
				else cr=FALSE;
				}
			}
		filecount++;
		}
    nsec=d64[blokk][1];                        /* stepping on the dir track */
	}
  while((ntrk=d64[blokk][0])!=0 && filecount < 144);
  if (cr) printf("\n");

  j=0; blokk=d64seek(18,0);                          /* BAM for free blocks */
  for (i=0;i<35;i++) {
	if (i!=17) j+=d64[blokk][4+i*4];
	}
  if (lowercase) printf ("%d blocks free.\n\n",j);
	else printf("%d BLOCKS FREE.\n\n",j);
  /* return(0); */
  }
}

/*** printing help and exit program ***/
void helpexit() {
    printf("%s\n","D64 lister and extractor V0.2 by Lion of Chromance");
    printf("%s\n","Usage: d64list [-lcef] [diskname.d64] ...");
	printf("%s\n","       -l = use lowercase charset");
	printf("%s\n","       -c = copy mode");
	printf("%s\n","       -e = add extensions in copy mode");
	printf("%s\n","       -f = print input filenames");
	exit(1);
}

int main (int argc,char *argv[]) {
  char *s;

  while (--argc>0 && (*++argv)[0]=='-')         /* commandline parameters */
	for (s=argv[0]+1;*s!='\0';s++)
		switch (*s) {
		case 'c': copymode=TRUE; break;
		case 'l': lowercase=TRUE; break;
        case 'e': fext=FALSE; break;
		case 'f': namepr=TRUE; break;
        case 'h':
        case '?': helpexit();

        default: printf("d64list:Illegal option %c\n",*s); helpexit(); break;
		} 
  if (argc < 1) {          /* no file, reading standart input (hi MEpk :)  */
    d64list(NULL);
	}
  else {                                   /* listing the files one by one */
    while (argc-- > 0) {
        d64list(*argv++);
        }
    }
}

